Sources/SwiftSyntax/SyntaxText.swift

```diff
@@ -0,0 +1,226 @@
+//===---------- SyntaxText.swift - Unowned String Representation  ---------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#if canImport(Darwin)
+import Darwin
+#elseif canImport(Glibc)
+import Glibc
+#endif
+
+/// Represent a string.
+///
+/// This type does not own the string data. The data reside in some other buffer
+/// whose lifetime extends past that of the SyntaxText.
+///
+/// `SyntaxText` conforms to `Collection` where the element is `UInt8`.
+@_spi(Testing) // SPI name is subject to change
+public struct SyntaxText {
+  var buffer: UnsafeBufferPointer<UInt8>
+
+  public init(baseAddress: UnsafePointer<UInt8>?, count: Int) {
+    assert(count == 0 || baseAddress != nil,
+           "If count is not zero, base address must be exist")
+    buffer = .init(start: baseAddress, count: count)
+  }
+
+  /// Creates an empty `SyntaxText`
+  public init() {
+    self.init(baseAddress: nil, count: 0)
+  }
+
+  /// Creates a `SyntaxText` from a `StaticString`
+  public init(_ string: StaticString) {
+    self.init(baseAddress: string.utf8Start, count: string.utf8CodeUnitCount)
+  }
+
+  /// Base address of the memory range this string refers to.
+  public var baseAddress: UnsafePointer<UInt8>? {
+    buffer.baseAddress
+  }
+
+  /// Byte length of this string.
+  public var count: Int {
+    buffer.count
+  }
+
+  /// A Boolean value indicating whether a string has no characters.
+  public var isEmpty: Bool {
+    buffer.isEmpty
+  }
+
+  /// Returns `true` if the memory range of this string is a part of `other`.
+  ///
+  /// `text[n ..< m].isSliceOf(text)` is always true as long as `n` and `m` are
+  /// valid indices.
+  public func isSliceOf(_ other: SyntaxText) -> Bool {
+    guard !isEmpty && !other.isEmpty else {
+      return isEmpty == other.isEmpty
+    }
+    return (other.baseAddress! <= baseAddress! &&
+            baseAddress! + count <= other.baseAddress! + other.count)
+  }
+
+  /// Returns `true` if `other` is a substring of this `SyntaxText`.
+  ///
+  /// This should not be used because `SyntaxText` is a `Collection<UInt8>` and
+  /// `contains(_:)` is usually for checking a collection contains the element.
+  /// Using `contains(_:)` for subsequence is confusing.
+  @available(*, deprecated, message: "Use 'firstRange(of:) != nil'")
```

Comment:
Why do we need this function if we just deprecate it straight away?

Reply:
I added this because `llvm::StringRef` has `contains(StringRef Other)` method and I feel that is a common operation. When C++ compiler contributor starts to work on this project, they would expect `SyntaxText` has `contains(_: Self)` method.

Reply:
I see. I’m not sure if it makes too much sense now anymore though, as the type is no longer named  `StringRef`. I think this should either be a properly supported method or removed.

Reply:
Oh I was wrong. [Swift 5.7 is actually adding `contains(_: Collection<Self.Element>)`](https://github.com/apple/swift-experimental-string-processing/blob/e47cc63a0a99b76ced6279292fe30db3979f4ff0/Sources/_StringProcessing/Algorithms/Algorithms/Contains.swift#L24-L37). So I'm going to remove `@available` here and leave this as a specialized method.